### ISSUE TITLE
questions about pixel-f1

### ISSUE THREAD

**Issue by chchshshhh** [dong03/MVSS-Net#7]
Hi,
Thank you very much for your work. I read your paper and tested your code (casia1.0) these days, but there are some doubts about the pixel F1 score
As shown in the following image,In the case of fixed threshold , the image F1 and pixel f1 are consistent, but  I want to get the best threshold test results,What can I do to get it? thanks
![image](https://user-images.githubusercontent.com/31399261/132156675-d700a388-b9c3-495c-bb43-dc5cfff8a8e4.png)

![image](https://user-images.githubusercontent.com/31399261/132159351-09e2e7c0-9c5f-43b5-8b7d-977fbc711439.png)

![image](https://user-images.githubusercontent.com/31399261/132160797-1393b28c-4874-4830-b549-6bc74ee71419.png)

![image](https://user-images.githubusercontent.com/31399261/132160734-59faa05d-fb61-4815-9145-eb56d09f6bd8.png)

**Comment by dong03** [comment]
You may refer to MFCN(Image Splicing Localization using a Multi-task Fully Convolutional Network).

However, why not using fixed threshold? In a pratical situation, you do not have a groundtruth to adjust threshold. 
0.5 is the most practical, reasonable and commonly adopted threshold.

**Comment by chchshshhh** [comment]
I want to see this because of my own evaluation (mask=Image.open(mask).convert('1') conversion and then use NEAREST interpolation to make the mask 512*512, and then compare it with the actual prediction map) The threshold is also 0.5. The evaluation is done by putting each point of each batch into the confusion matrix, and then after all the samples, the calculation of f1 is performed, but it is strange that the mvss model obtained by my own training (no classification only segmentation) is in yours F1 is 0.45 under the evaluation, but it is 0.57 on my evaluation f1, but your pre-trained model is 0.61 on my evaluation f1, which are all pixel-level evaluations. I didn’t understand. The evaluation method is as follows
My training：
![image](https://user-images.githubusercontent.com/31399261/132870257-8af5280b-7517-4bf1-8fea-cb5f277f2461.png)
my training is based on my method：
![image](https://user-images.githubusercontent.com/31399261/132878872-a3eef9a9-de4f-4d0f-8572-79224b74e413.png)
Your pre-training is based on my evaluation method：
![image](https://user-images.githubusercontent.com/31399261/132879014-bb74bac7-b52a-4d5c-b53c-d3ee51655601.png)
my evaluation method:
![image](https://user-images.githubusercontent.com/31399261/132880468-363c6258-f587-4561-a40a-d1fef1cc40a0.png)

**Comment by Chenxr1999** [comment]
We notice that there is a gap between our evaluation process. According to your code, you sum up the confusion matrix of multiple images, and get one f1 value. While in our code and the statement in the paper, we calculate f1 of every single test image separately, and then take the average value of all f1s as the pixel-level f1. Besides, we resize the 512*512 prediction map into the origin size, while you resize the mask into 512*512, which may also lead to a small gap between the evaluation result.

**Comment by chchshshhh** [comment]
I thought about the next two evaluation methods and talked about my understanding. My method is actually to do f1 for all pixels, showing the extent of the pixels. For example, all pixels in a single image are predicted wrong, and a single image f1 is 0. The average of all images in your assessment may have a relatively large impact, but mine will reduce this difference in the case of particularly multi-pixels. The same is true for all predictions. My personal understanding is this, thank you very much for your reply in the end

**Comment by 85zhanghao** [comment]
> I thought about the next two evaluation methods and talked about my understanding. My method is actually to do f1 for all pixels, showing the extent of the pixels. For example, all pixels in a single image are predicted wrong, and a single image f1 is 0. The average of all images in your assessment may have a relatively large impact, but mine will reduce this difference in the case of particularly multi-pixels. The same is true for all predictions. My personal understanding is this, thank you very much for your reply in the end



> I thought about the next two evaluation methods and talked about my understanding. My method is actually to do f1 for all pixels, showing the extent of the pixels. For example, all pixels in a single image are predicted wrong, and a single image f1 is 0. The average of all images in your assessment may have a relatively large impact, but mine will reduce this difference in the case of particularly multi-pixels. The same is true for all predictions. My personal understanding is this, thank you very much for your reply in the end

Do you get the best F1 score (75.3)  in the case of optimal threshold?Can you tell  me  how  to get the  optimal threshold? Thank you !

**Comment by 85zhanghao** [comment]
Do you get the best F1 score (75.3) in the case of optimal threshold?Can you tell me how to get the optimal threshold? Thanks!

**Comment by chchshshhh** [comment]
> Do you get the best F1 score (75.3) in the case of optimal threshold?Can you tell me how to get the optimal threshold? Thanks!

I didn't achieve the same score, I just followed the published optimal threshold method (constrain-rcnn) for model comparison.

**Comment by 85zhanghao** [comment]
> > Do you get the best F1 score (75.3) in the case of optimal threshold?Can you tell me how to get the optimal threshold? Thanks!
> 
> I didn't achieve the same score, I just followed the published optimal threshold method (constrain-rcnn) for model comparison.
 I also  followed the optimal threshold method in  constrain-rcnn. The  F1 score I got  is  62.7, what about you?

**Comment by chchshshhh** [comment]
> > > Do you get the best F1 score (75.3) in the case of optimal threshold?Can you tell me how to get the optimal threshold? Thanks!
> > 
> > 
> > I didn't achieve the same score, I just followed the published optimal threshold method (constrain-rcnn) for model comparison.
> > I also  followed the optimal threshold method in  constrain-rcnn. The  F1 score I got  is  62.7, what about you?

It's the same for me. In fact, if there are background pixels, I can also achieve F1 scores of 70 and above, but it does not match the results of this paper, and it's not fair compared to the previous optimal threshold paper, so I used F1 score of 62.7 as a benchmark to compare other papers.

**Comment by 85zhanghao** [comment]
> > > > Do you get the best F1 score (75.3) in the case of optimal threshold?Can you tell me how to get the optimal threshold? Thanks!
> > > 
> > > 
> > > I didn't achieve the same score, I just followed the published optimal threshold method (constrain-rcnn) for model comparison.
> > > I also  followed the optimal threshold method in  constrain-rcnn. The  F1 score I got  is  62.7, what about you?
> 
> It's the same for me. In fact, if there are background pixels, I can also achieve F1 scores of 70 and above, but it does not match the results of this paper, and it's not fair compared to the previous optimal threshold paper, so I used F1 score of 62.7 as a benchmark to compare other papers.

what is  the background pixels mean？I'm not sure. Can you explain about it?
